lib/puppet/provider/foreman_resource/rest_v3.rb

```diff
@@ -95,6 +98,20 @@ def success?(response)
   end
 
   def error_message(response)
-    JSON.parse(response.body)['error']['full_messages'].join(' ') rescue "unknown error (response #{response.code})"
+    messages = {
+      '400' => '400 Bad Request: Something is wrong with the data we sent to Foreman server',
+      '401' => '401 Unauthorized Request: Check credentials for validity',
```

Comment:
Can we call out what credentials to check? The oauth credentials? Or the client ceritificates?

Reply:
Yes, I agree this is a bit of a copout.

I was able to reproduce the 401 response by introducing a typo in the `--foreman-proxy-oauth-consumer-key`. Another method to reproduce it is to introduce a typo in the `--foreman-proxy-foreman-base-url`. Using incorrect certificates should result in the same response (though I have not personally tested it).

While I don't insist that we have to implement it right this minute, I think this illustrates the use case for branded error messages. We could link here to a community wiki with detailed information on how to troubleshoot the error, while downstream could link to a KCS describing the error and troubleshooting steps.

Reply:
The only authentication that the Foreman server does is oauth credentials. AFAIK no client certificates are used so the only certificate is the server CA that's verified. If that's invalid, no HTTP(S) connection can be established in the first place. However, if they point it at a different server/service then who knows. There may also be some HTTP proxy that does odd things. Because of that last point I'm hesitant in trying to explain exactly what's going on.

Reply:
The next version I push will contain the explanation `'Usually this indicates something was wrong with the Oauth credentials'`